Added support for pagination to github_deploy_key

bincyber opened the pull request:
##### SUMMARY

Added support for pagination to the `github_deploy_key` module, so that it works when more than 30 deploy keys exist for a repository.

I needed to extract links from the HTTP link header, so I added `parse_header_links` function to extract links from the HTTP Link header. The code was taken from the requests library with slight modification: https://github.com/requests/requests/blob/master/requests/utils.py#L812

Removed json import in place of `module.from_json()` and `module.jsonify()`.

Refactored module.

Fixes: #36745

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
github_deploy_key module

##### ANSIBLE VERSION

```
ansible 2.4.3.0
  config file = None
  configured module search path = [u'/home/vagrant/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, Aug  4 2017, 00:39:18) [GCC 4.8.5 20150623 (Red Hat 4.8.5-16)]
```


##### ADDITIONAL INFORMATION

bincyber committed:
Fixes #36745

- added support for pagination to the github_deploy_key module
- removed json import
- removed requirements from documentation
- refactored the module
- added changelog fragment

punkrokk commented:
@bincyber am I out of line recommending you not build the API link, but always take the next from the API? GH recommends that, as in some cases the pagination could change mid request. Probably unlikely with small numbers of deploy keys, but possible when you are hovering over a value mod 30. 

From here: https://developer.github.com/v3/guides/traversing-with-pagination/
Always rely on these link relations provided to you. Don't try to guess or construct your own URL.

bincyber commented:
@punkrokk No, that's a great suggestion. I have modified the way pagination is handled so it always uses the `next` URL in the Link header instead of constructing it myself. I've also refactored the code a bit.

punkrokk commented:
Yea that looks great. Can’t wait to try it. Mind explaining your use of yield on line 212?

bincyber commented:
@punkrokk `paginate()` is a [generator](https://www.programiz.com/python-programming/generator) used to loop through the Link header. `yield` returns the deploy keys for each page

ready_for_review

bincyber commented:
This PR has been open for months. Can we get this merged?

punkrokk commented:
@bincyber Hey how can we get this merged ?

bincyber commented:
@gundalow can we please get this bugfix merged?

gundalow commented:
Needs a rebase then can be merged.

I'm on parental leave at the moment, so once it's been rebase please mention this in `#ansible-devel` on Freenode IRC and someone will hit merged.

bincyber commented:
Rebased.

@punkrokk can you give this one final test before I request @Akasurde to review and merge?

gundalow commented:
@punkrokk would you be able to review this please?

punkrokk commented:
Yes I’ll look today or Monday

Cheers,

JP

> On Oct 18, 2019, at 8:45 AM, John R Barker <notifications@github.com> wrote:
> 
> ﻿
> @punkrokk would you be able to review this please?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

punkrokk reviewed (approved):
I was able to check this out and this worked the same as I expected and when I test it awhile ago. No concerns with the code.

flowerysong commented:
I have concerns about the source code provenance; requests is distributed under the Apache license, while module_utils/urls.py is BSD. It's fairly simple code and a reimplementation from scratch might have turned out looking exactly like it does now, but explicitly basing it on a library with a different license is very iffy.

felixfontein commented:
1. This needs a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#changelogs-how-to).
2. About the requests issue: I added a function parsing the links header to module_utils/acme.py some time ago: https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/acme.py#L959-L966 Feel free to move it to module_utils/urls.py (the modules using it are in lib/ansible/modules/crypto/acme) and use it instead.

punkrokk commented:
@flowerysong you have a good point here. @bincyber how do you want to proceed?

bincyber commented:
Thanks for the feedback all. I've removed the code copied from `requests` and copied the regex from [process_links](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/acme.py#L975).

I've also added the changelog fragment and rebased ontop of devel branch. Let's get this merged!

gundalow commented:
Thanks, if people could please review and add `shipit` comments I'll get this merged.

Akasurde commented:
@bincyber Thanks for the contribution. @punkrokk @gundalow @flowerysong @felixfontein Thanks for the review.
